src/fs/metadata_log_bootstrap.hh

```diff
@@ -0,0 +1,485 @@
+        if (not std::holds_alternative<inode_info::file>(_metadata_log._inodes[entry.entry_inode].contents)) {
+            return invalid_entry_exception();
+        }
+
+        if (not std::holds_alternative<inode_info::directory>(_metadata_log._inodes[entry.dir_inode].contents)) {
+            return invalid_entry_exception();
+        }
+        auto& dir = std::get<inode_info::directory>(_metadata_log._inodes[entry.dir_inode].contents);
+
+        if (dir.entries.count(dir_entry_name) != 0) {
+            return invalid_entry_exception();
+        }
+
+        _metadata_log.memory_only_add_dir_entry(dir, entry.entry_inode, std::move(dir_entry_name));
+        // TODO: Maybe mtime_ns for modifying directory?
+        return now();
+    }
+
+    future<> bootstrap_create_inode_as_dir_entry() {
+        ondisk_create_inode_as_dir_entry_header entry;
+        if (not _curr_checkpoint.read_entry(entry) or not inode_exists(entry.dir_inode) or inode_exists(entry.entry_inode.inode)) {
+            return invalid_entry_exception();
+        }
+
+        sstring dir_entry_name;
+        if (not _curr_checkpoint.read_sstring(dir_entry_name, entry.entry_name_length)) {
+            return invalid_entry_exception();
+        }
+
+        if (not std::holds_alternative<inode_info::directory>(_metadata_log._inodes[entry.dir_inode].contents)) {
+            return invalid_entry_exception();
+        }
+        auto& dir = std::get<inode_info::directory>(_metadata_log._inodes[entry.dir_inode].contents);
+
+        if (dir.entries.count(dir_entry_name) != 0) {
+            return invalid_entry_exception();
+        }
+
+        _metadata_log.memory_only_create_inode(entry.entry_inode.inode, entry.entry_inode.is_directory, ondisk_metadata_to_metadata(entry.entry_inode.metadata));
+        _metadata_log.memory_only_add_dir_entry(dir, entry.entry_inode.inode, std::move(dir_entry_name));
+        // TODO: Maybe mtime_ns for modifying directory?
+        return now();
+    }
+
+    future<> bootstrap_delete_dir_entry() {
+        ondisk_delete_dir_entry_header entry;
+        if (not _curr_checkpoint.read_entry(entry) or not inode_exists(entry.dir_inode)) {
+            return invalid_entry_exception();
+        }
+
+        sstring dir_entry_name;
+        if (not _curr_checkpoint.read_sstring(dir_entry_name, entry.entry_name_length)) {
+            return invalid_entry_exception();
+        }
+
+        if (not std::holds_alternative<inode_info::directory>(_metadata_log._inodes[entry.dir_inode].contents)) {
+            return invalid_entry_exception();
+        }
+        auto& dir = std::get<inode_info::directory>(_metadata_log._inodes[entry.dir_inode].contents);
+
+        auto it = dir.entries.find(dir_entry_name);
+        if (it == dir.entries.end()) {
+            return invalid_entry_exception();
+        }
+
+        if (_metadata_log._inodes.at(it->second).directories_containing_file == 0) {
```

Comment:
It looks like it should be 'assert' instead of 'if'.

Reply:
Right. That assert is already in memory_only_delete_dir_entry(), so I will remove this if completely.